Source/Core/Common/DirectIOFile.cpp

```diff
@@ -0,0 +1,363 @@
+{
+  Swap(other);
+}
+
+DirectIOFile& DirectIOFile::operator=(DirectIOFile&& other)
+{
+  Close();
+  Swap(other);
+  return *this;
+}
+
+DirectIOFile::DirectIOFile(const std::string& path, AccessMode access_mode, OpenMode open_mode)
+{
+  Open(path, access_mode, open_mode);
+}
+
+bool DirectIOFile::Open(const std::string& path, AccessMode access_mode, OpenMode open_mode)
+{
+  ASSERT(!IsOpen());
+
+  if (open_mode == OpenMode::Default)
+    open_mode = (access_mode == AccessMode::Write) ? OpenMode::Truncate : OpenMode::Existing;
+
+  // This is not a sensible combination. Fail here to not rely on OS-specific behaviors.
+  if (access_mode == AccessMode::Read && open_mode == OpenMode::Truncate)
+    return false;
+
+#if defined(_WIN32)
+  DWORD desired_access = GENERIC_READ | GENERIC_WRITE;
+  if (access_mode == AccessMode::Read)
+    desired_access = GENERIC_READ;
+  else if (access_mode == AccessMode::Write)
+    desired_access = GENERIC_WRITE;
+
+  // Allow deleting and renaming through our handle.
+  desired_access |= DELETE;
+
+  // All sharing is allowed to more closely match default behavior on other OSes.
+  constexpr DWORD share_mode = FILE_SHARE_DELETE | FILE_SHARE_READ | FILE_SHARE_WRITE;
+
+  DWORD creation_disposition = OPEN_ALWAYS;
+  if (open_mode == OpenMode::Truncate)
+    creation_disposition = CREATE_ALWAYS;
+  else if (open_mode == OpenMode::Create)
+    creation_disposition = CREATE_NEW;
+  else if (open_mode == OpenMode::Existing)
+    creation_disposition = OPEN_EXISTING;
+
+  m_handle = CreateFile(UTF8ToTStr(path).c_str(), desired_access, share_mode, nullptr,
+                        creation_disposition, FILE_ATTRIBUTE_NORMAL, nullptr);
+  if (!IsOpen())
+    WARN_LOG_FMT(COMMON, "CreateFile: {}", Common::GetLastErrorString());
+
+#elif defined(ANDROID)
+  // Leveraging IOFile to avoid reimplementing things.
+  // A few features are emulated in a non-atomic manner.
+
+  // Allow write without truncate. This wrongly enables read access.
+  if (access_mode == AccessMode::Write && open_mode != OpenMode::Truncate)
+    access_mode = AccessMode::ReadAndWrite;
+
+  // FYI: File::Exists can be slow on Android.
+  Common::Lazy<bool> file_exists{[&] { return Exists(path); }};
+
+  // Require a new file be created.
+  if (open_mode == OpenMode::Create && *file_exists)
+    return false;
+
+  // Note: `Existing` doesn't need a manual test because we only use "w" with `Truncate`.
+
+  // We must manually create a file if we don't use "w".
+  if (open_mode != OpenMode::Existing && access_mode != AccessMode::Write && !*file_exists)
+    CreateEmptyFile(path);
+
+  const char* open_mode_str = "rb+";
+  if (access_mode == AccessMode::Read)
+    open_mode_str = "rb";
+  else if (access_mode == AccessMode::Write)
+    open_mode_str = "wb";
+
+  if (IOFile file(path, open_mode_str); file.IsOpen())
+  {
+    // Make `OpenMode::Truncate` work with `AccessMode::ReadAndWrite`.
+    // Also, apparently "w" may or may not truncate on Android..
```

Comment:
Actually, if you pass "w" to IOFile, you will get truncation. This is because IOFile calls this function for SAF paths:

https://github.com/dolphin-emu/dolphin/blob/9d27c145b9a43d58f4a2827d60626357aba428af/Source/Android/jni/AndroidCommon/AndroidCommon.cpp#L62-L82

It would probably be beneficial to inline the logic of `IOFile::Open` into `DirectIOFile::Open` (i.e. calling AndroidCommon.cpp from here instead of creating an `IOFile`) to make all this logic less "magic". Note also how `IOFile::Open` falls back to the regular Unix code if `IsPathAndroidContent` returns false.

Reply:
In theory anything could happen, since any app is able to write a storage provider that does some complete nonsense if they want to. But if we don't assume anything at all, there's literally nothing we can use, so I would say it's fine to assume that "r" and "rw" don't truncate. On the other hand, we shouldn't assume anything about "w", since Google specifically says that one is ambiguous. Therefore, I think it would be best to re-add the logic of using "rw" in place of non-truncating "w". (Or can we use "wa"?)

Alternatively: Just assert if someone wants to do a non-truncating write. I don't think we have any code that's doing that, and it would be far from the only file operation that we don't support under SAF.